designs/default-zero-value.md

```diff
@@ -0,0 +1,857 @@
+
+
+#### `aws.api#clientOptional` trait
+
+The `aws.api#clientOptional` trait is used to indicate that a structure member
+SHOULD be unconditionally generated as optional regardless of if the member
+targets a shape with a default value, or if the member is marked with th
+ `@default` trait, or if the member is marked with the `@required` trait.
+This trait allows documentation generators to indicate that a member is
+required, even if it is not reflected in generated code.
+
+The `aws.api#clientOptional` trait is defined in Smithy as:
+
+```
+$version: "2"
+namespace aws.api
+
+@trait(selector: "structure > member")
+structure clientOptional {}
+```
+
+Consider the following model:
+
+```
+$version: "2"
+namespace smithy.examnple
+
+use aws.api#clientOptional
+
+structure ProductData {
+    @clientOptional
+    @required
+    description: String
+}
+```
+
+When generating an AWS SDK client for this shape, the `ProductData$description`
+member MUST be generated as an optional member rather than always present.
+
+
+##### Backfilling the clientOptional trait
+
+The `aws.api#clientOptional` trait will be backfilled onto AWS models as
+needed. Some more problematic services will use this trait with every required
+structure member, whereas others will only use it on structure members that
+target structure or union shapes that have no default zero value. Over time, as
+models are audited and corrected, we can remove the `@clientOptional` trait
+and release improved AWS SDKs.
+
+Backfilling this trait on AWS models is admittedly be an inconvenience for
+developers, but it is less catastrophic than previously generated client code
+failing at runtime when deserializing the response of an operation, and it
+matches the current reality of how AWS was modeled. While it's true client code
+might start failing anyways if a service stops sending output members that were
+previously marked as required, it would only fail when the client explicitly
+attempts to dereference the value, and this kind of change is generally only
+made when a client opts-into some new kind of functionality or workflow. If an
+SDK fails during deserialization because a previously required output member
+is missing, then customer applications would be completely broken until they
+update their SDK, regardless of if the member is used in client code.
+
+
+## Alternatives and trade-offs
+
+## A note on added complexity in Smithy 1.x
+
+Nullability rules in Smithy 1.x were already complex, and we're making them more
+complex in this proposal. Before this proposal, a member was considered nullable
+if it was marked with the `@box` trait or if it targeted a shape marked with the
+`@box` trait, or anything other than a boolean or number. With this proposal, we
+are also introducing the `@default` trait, further complicating nullability. We
+will address this added complexity if we release a Smithy 2.x that removes the
```

Comment:
This sentence is out of date with the proposed IDL 2.0 that does remove the `@box` trait.